### PR TITLE
SinkIslandCheck Enhancements to Exclude Airport Roads

### PR DESCRIPTION
**Pull request by sayas01**
### Description:

This PR adds the following enhancements to SinkIslandCheck:
1) Exclude all service roads that are within any airport polygons. Current version has a validation that excludes edges that have aeroway tags with values runway or taxiway. Since ways with aeroway tags are not ingested as edges, this validation has been removed and a method that checks if the road(service highways) is within airport polygon has been added.
2) Previously only the end node of the edge was checked for _amenity=parking_entrance_ which resulted in some false positives. Additional validation to check for _amenity=parking_entrance_ tag at the start node of the edge has been added which should fix these cases.
3) Service roads that ends in buildings have been excluded from the check.
4) Previously, navigability of the edge was checked only in terms of highway tags. Additional validation in terms of _Access_ tag and _Motorvehicle_ and _Motorcar_ tags have been added to determine the car navigability of the edge.

### Potential Impact:

From the analysis, it was seen that there is both addition and subtraction of flags with these enhancements. Subtractions have been verified to be the cases that should be removed by these enhancements like services roads within airports. Additions have been verified to be mainly due to the updated car navigability condition. Most of the additions are for edges that are not marked _access=no_ but are connected to _access=no_ edges, which are valid sink islands.

### Unit Test Approach:

Unit tests have been added for the new updates. Additional validation has been done by testing the updates on a sample of countries and verifying it with the previous version.

### Test Results:

All tests passing and the additions and subtractions have been verified to be valid.

### CODE REVIEW

File: src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java [hunk 1]
```diff
@@ -278,4 +285,37 @@ private boolean isConnectedToPedestrianNavigableHighway(final Edge edge)
...
+
+    /**
+     * Checks if the edge is car navigable or not. A car navigable edge is an edge with highway type
+     * within the car navigable enum set and with certain accessibility tag value = !no
+     *
+     * @param edge
+     *            any edge
+     * @return true if the edge is car navigable, else false
+     */
+    private boolean isCarNavigableHighway(final Edge edge)
```
Comment (danielduhh) [inline src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java]: Have you considered moving this to a location (`TypePredicates`) where other checks can pick this up?
Also, is motor_vehicle=yes required to use motor_vehicle=no to determine navigability ?

File: src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java [hunk 2]
```diff
@@ -224,14 +228,17 @@ private boolean edgeCharacteristicsToIgnore(final Edge edge)
...
+                || this.endOrStartNodeHasAmenityTypeToExclude(edge)
                 // Ignore edges that have been way sectioned at the border, as has high probability
                 // of creating a false positive due to the sectioning of the way
                 || SyntheticBoundaryNodeTag.isBoundaryNode(edge.end())
                 || SyntheticBoundaryNodeTag.isBoundaryNode(edge.start())
                 // Ignore edges that are of type service and is connected to pedestrian navigable
                 // ways
-                || this.isServiceRoad(edge) && this.isConnectedToPedestrianNavigableHighway(edge);
+                || this.isServiceRoad(edge) && this.isConnectedToPedestrianNavigableHighway(edge)
+                // Ignore service roads that ends in a building or are within any airport polygon
```
Comment (danielduhh) [inline src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java]: Why only service roads?

File: src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java (hunk 1 above)
Comment (sayas01) [reply src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java]: Having this method as a utility function in TypePredicate will be very useful. I did not go for it this time because it requires some deep dive into the definition of navigability. In case of SinkIsland, Keahi and I decided to go with "!motor_vehicle=no" because of the amount of flags that would be introduced by "!motor_vehicle=yes" and also because the requirement specifically cited FPs with "motorvehicle=no" tags. But ideally, "motorvehicle=yes" should determine the navigability of an edge.

File: src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java (hunk 2 above)
Comment (sayas01) [reply src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java]: The requirement was to remove service roads within airports. Also based on Keahi's analysis, service roads that end up in Buildings are valid edges.

File: src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java (hunk 1 above)
Comment (sayas01) [reply src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java]: @danielduhh I have updated to use a more generic tag predicate in: https://github.com/sayas01/atlas-checks/pull/12/commits/24afba08d193c639b3f7e7385a3befaf9c0bca09

File: src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java [hunk 3]
```diff
@@ -78,7 +82,9 @@ public boolean validCheckForObject(final AtlasObject object)
                 && ((Edge) object).highwayTag()
                         .isMoreImportantThanOrEqualTo(this.minimumHighwayType)
                 && !(this.isServiceRoad((Edge) object)
-                        && this.isWithinAreasWithExcludedAmenityTags((Edge) object));
+                        && this.isWithinAreasWithExcludedAmenityTags((Edge) object))
+                && !(Validators.isOfType(object, HighwayTag.class, HighwayTag.SERVICE)
```
Comment (Bentleysb) [inline src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java]: This seems somewhat redundant with the `this.isServiceRoad()` call above.

File: src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java [hunk 4]
```diff
@@ -278,4 +285,38 @@ private boolean isConnectedToPedestrianNavigableHighway(final Edge edge)
...
+     * with {@link AerowayTag} tag.
+     *
+     * @param edge
+     *            any edge
+     * @return true if the edge is within or intersects airport polygon
+     */
+    private boolean intersectsAirportOrBuilding(final Edge edge)
+    {
+        return StreamSupport.stream(edge.getAtlas()
+                .areasIntersecting(edge.bounds(),
```
Comment (Bentleysb) [inline src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java]: With using the `edge.bounds()` for the intersect there is a possibility of false negatives just outside airports, right? Albeit, the chance is probably quite low.

File: src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java (hunk 1 above)
Comment (Bentleysb) [reply src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java]: There are definite intricacies to navigability in OSM. In this case it looks like the largest piece you are missing is motor_vehicle/motorcar/vehicle=yes overriding access=no. I think it is definitely worth researching and writing out a utility function for this.

File: src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java (hunk 3 above)
Comment (sayas01) [reply src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java]: In case of roads that are within airports or that ends in buildings, we are considering all service roads and not just the ones that have service tag along with highway=service. I could refactor the code to use the validations in `isServiceRoad()` in line if the method call is confusing.

Comment (Bentleysb) [reply src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java]: Why do the uses of isServiceRoad require there to be a service tag?

Comment (sayas01) [reply src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java]: I think that was part of the previous enhancements were the use cases required the roads to be having both highway=service and service=* tags.

Comment (Bentleysb) [reply src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java]: Ok, it could be worth reviewing those cases to see if we can combine things. Assuming this is the way it needs to be, though, it would be worth added a comment or two about this.

File: src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java (hunk 1 above)
Comment (sayas01) [reply src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java]: @kkonishi2 what do you think about updating the car navigability validation for this check?

File: src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java (hunk 4 above)
Comment (sayas01) [reply src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java]: I would say the probability would be very low. One other way would be to grab all areas containing the nodes of the edge matching a predicate. What do you think?

Comment (Bentleysb) [reply src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java]: Check this out for a solution I used a long while ago:
https://github.com/osmlab/atlas-checks/blob/dev/src/main/java/org/openstreetmap/atlas/checks/validation/tag/InvalidAccessTagCheck.java#L112

However, in this case it may be so unlikely that it is not worth the extra processing.

File: src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java [hunk 5]
```diff
@@ -183,30 +189,28 @@ public boolean validCheckForObject(final AtlasObject object)
...
      */
-    private boolean endNodeHasAmenityTypeToExclude(final AtlasObject object)
+    private boolean endOrStartNodeHasAmenityTypeToExclude(final AtlasObject object)
     {
         final Edge edge = (Edge) object;
         final Node endNode = edge.end();
-
-        return Validators.isOfType(endNode, AmenityTag.class, AMENITY_VALUES_TO_EXCLUDE);
+        return Validators.isOfType(endNode, AmenityTag.class, AMENITY_VALUES_TO_EXCLUDE)
+                || Validators.isOfType(edge.start(), AmenityTag.class, AmenityTag.PARKING_ENTRANCE);
```
Comment (nachtm) [inline src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java:221]: nit: could we replace `endNode` with `edge.end()` here? Or add `final Node startNode = edge.start()`? Just make the two lines more similar for readability

File: src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java (hunk 1 above)
Comment (nachtm) [reply src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java]: It seems like there's a need for a unified definition of OSM navigability in this repo (or in atlas itself). 

The research involved in doing that properly might be outside the scope of this check, but I would vote that we look into doing the necessary research. We could either say that this PR is blocked until that change happens, or we could push this through and refactor it to use the navigability definition later. Either way is fine with me -- what do you all think?

File: src/main/java/org/openstreetmap/atlas/checks/atlas/predicates/TagPredicates.java [hunk 6]
```diff
@@ -106,4 +107,17 @@
      */
     Predicate<AtlasObject> IS_CROSSING_HIGHWAY = object -> Validators.isOfType(object,
             HighwayTag.class, HighwayTag.CROSSING);
+
+    /**
+     * Tests if the {@link AtlasObject} has car navigable highway type and access tag not in Private
+     * enum set
+     */
+    Predicate<AtlasObject> IS_CAR_NAVIGABLE_NON_PRIVATE_HIGHWAY = object -> HighwayTag
+            .isCarNavigableHighway(object) && !AccessTag.isPrivate(object);
```
Comment (nachtm) [inline src/main/java/org/openstreetmap/atlas/checks/atlas/predicates/TagPredicates.java]: I think if we're going to add something to a more general location like this, we should do the research to verify that it's the correct OSM definition.

File: src/test/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheckTest.java [hunk 7]
```diff
@@ -83,7 +85,7 @@ public void testInvalidEdges()
         this.verifier.actual(this.setup.getInvalidEdges(),
                 new SinkIslandCheck(ConfigurationResolver.emptyConfiguration()));
         this.verifier.verifyExpectedSize(1);
-        this.verifier.verify(flag -> Assert.assertEquals(2, flag.getFlaggedObjects().size()));
+        this.verifier.verify(flag -> Assert.assertEquals(3, flag.getFlaggedObjects().size()));
```
Comment (nachtm) [inline src/test/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheckTest.java:88]: Why the change here?

File: src/test/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheckTestRule.java [hunk 8]
```diff
@@ -116,7 +125,7 @@
             @Node(coordinates = @Loc(value = TEST_2)),
             @Node(coordinates = @Loc(value = TEST_3)) }, edges = {
                     @Edge(coordinates = { @Loc(value = TEST_1), @Loc(value = TEST_2) }, tags = {
-                            "highway=service", "aeroway=taxiway" }),
+                            "highway=service" }),
```
Comment (nachtm) [inline src/test/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheckTestRule.java]: ...and the change here?

**Review by nachtm — COMMENTED** [review]
Don't forget to add an `@author` tag!

File: src/test/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheckTestRule.java (hunk 8 above)
Comment (sayas01) [reply src/test/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheckTestRule.java]: Since ways with "aeroway" tags are ingested as lines, I removed this tag from the test data for this edge.

File: src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java (hunk 1 above)
Comment (danielduhh) [reply src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java]: check************ @nachtm

File: src/test/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheckTest.java (hunk 7 above)
Comment (sayas01) [reply src/test/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheckTest.java:88]: Previously, edge with aeroway tag was excluded from the check. But since aeroway tagged ways are not ingested as edges, that validation was removed and so this edge will be flagged as sink island.

Comment (nachtm) [reply src/test/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheckTest.java:88]: Ok. I'm a little confused about removing this validation -- was it creating false positives? Are we sure that the `line` definition won't change in the future?

Comment (sayas01) [reply src/test/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheckTest.java:88]: I think at some point aeroway ways were ingested as edges. Since these are not ingested as edges, the validation to remove aeroway ways from this check was not applied and so the edges with aeroways were still flagged as sink island. The new enhancement to exclude service roads within airports would capture these cases.

File: src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java (hunk 1 above)
Comment (kkonishi2) [reply src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java]: I think that the way we have it set up is fine. Simply looking for a motorway=yes tag may lead to unforeseen issues. For example, you're able to tag a Way as access=private and motor_vehicle=yes, potentially leading to some confusing situations. Were this not an open source check, I would be more concerned with trying to keep standards uniform, but because this is open source, I think it's pretty safe to treat Ways with tags such as access=customer as essentially non-navigable ways. Most routing engines will assign high costs to roads based on these tags, so these are essentially treated as non-navigable anyway.

One thing to consider is that in our current configuration, flag count has increased in some places fairly significantly because of the presence of navigable Ways within non-navigable networks (**image below**). I believe Identifying these for fix is a net positive to the map quality, not only from a navigability standpoint but also from an aesthetic one. Again, were this an open source check, I would be more concerned with things such as editor time/fix count, but because this is open source and the main goal is map quality, I think this is the way to go. 


Screen Shot 2019-06-26 at 11 19 10 AM

Comment (sayas01) [reply src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java]: Thanks @kkonishi2 ! @danielduhh What is your thoughts on the above comment? Having a shared navigability criteria in atlas/atlas-checks would be ideal. I could work with @kkonishi2 to come up with validations for a shared car navigability function. Doesn't look like there is a unified navigability criteria in OSM as per @kkonishi2.

Comment (danielduhh) [reply src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java]: Doesn't motor_vehicle/motorcar/vehicle=yes override access=no? The access tag is more general in terms of public access, whereas the transport mode tags are more car navigation specific

Comment (danielduhh) [reply src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java]: thoughts?
@Bentleysb @nachtm

Comment (nachtm) [reply src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java]: From what I understand, access=no and motor_vehicle=yes means only motor_vehicles can drive down a road. So yes, a specific navigation tag should override access=no. This is supported by [the wiki page on the access tag](https://wiki.openstreetmap.org/wiki/Key:access): 

>  - access=no, bus=yes means that only buses are allowed to enter (for example a road only for buses)
> - access=yes, motor_vehicle=no means that all transport modes except motor vehicles can use the element
> - access=forestry, foot=permissive implies that forestry vehicles can use the route legally and that pedestrians can use it by currently but that permission may later be withdrawn.

Comment (sayas01) [reply src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java]: I feel like we need to decide on whether we should flag :
1) `all accessible highways that are not private -> access=!no and motorvehicle/motorcar/vehicle=!no` or
2) `public highways that are accessible -> access!=private and motorvehicle/motorcar/vehicle=yes`

Comment (sayas01) [reply src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java]: I have updated the conditions for car navigability to be access not in private enum set(as per the analysis done on SinkIsland) and transportation tags=yes. Feel like this should capture all the car navigable cases and skip all private roads.

Comment (seancoulter) [reply src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java]: Micah's comment tells me that the access tag doesn't seem to be a reliable measure of actual accessibility. And I think a car navigable sink island is more problematic than a car non-navigable sink island because the former is impossible to enter/exit given their road network and despite access rights. So maybe we can flag ```motorvehicle/motorcar/vehicle=!no && isNotConnectedToCarNavigableEdge()``` and ignore the access tag.

File: src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java (hunk 4 above)
Comment (seancoulter) [reply src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java]: To add to this--just to be safe we could check if the airport intersects with the edge itself, not the edge's bounds. This could be accomplished using the ```overlaps()``` function of ```GeometricSurface```. Here's the method description for ```overlaps()``` in ```Polygon``` (Polygon implementing GeometricSurface):
```
     * Tests if this {@link Polygon} intersects/overlaps the given {@link PolyLine} at any point.
     * This is different than the {@link #fullyGeometricallyEncloses(PolyLine)} method, in that it
     * doesn't require full containment within the {@link Polygon}, just any overlap.
```
We could do something like ```AirportPolygon.overlaps(Edge)```. Internally it looks like this is comparing lat/longs of the concerned objects.

**Review by seancoulter — COMMENTED** [review]
I had a couple of comments re navigability tags and edges in airports/buildings. Everything else looks great!

File: src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java (hunk 1 above)
Comment (sayas01) [reply src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java]: @kkonishi2  and I went through the different combinations and we feel like SinkIslandCheck should only look at edges that have access = yes and motorvehicle/vehicle/motorcar=yes tags. There are private roads with _motorvehicle/vehicle/motorcar tag value = yes_, like roads within a college campus, which may not be connected to any other navigable edges. These private roads need to be flagged as SinkIslands.

File: src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java [hunk 9]
```diff
@@ -183,30 +194,31 @@ public boolean validCheckForObject(final AtlasObject object)
...
-                // Ignore any airport taxiways and runways, as these often create a sink island
-                && !Validators.isOfType(object, AerowayTag.class, AerowayTag.TAXIWAY,
-                        AerowayTag.RUNWAY)
-                // Only allow car navigable highways and ignore ferries
-                && HighwayTag.isCarNavigableHighway(object) && !RouteTag.isFerry(object)
+                // Only allow car navigable highways (access = yes and
+                // motorvehicle/motorcar/vehicle = yes)
+                // and ignore ferries
+                && HighwayTag.isCarNavigableHighway(object)
+                && Validators.isOfType(object, AccessTag.class, AccessTag.YES)
```
Comment (danielduhh) [inline src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java]: I'm not following why we added this. My first thought is this is way too strict and will drastically limit the number of flags. Am I missing something?

**Comment by sayas01** [comment]
@danielduhh @nachtm @Bentleysb @seancoulter Analysis of the flags generated using the updated car navigability condition (https://github.com/sayas01/atlas-checks/pull/12/commits/580ceb14a8497dbc5faf7dde732682b3db4e9e6e), showed there were some FPs due to missing transportation tags. This [road](https://www.openstreetmap.org/way/285163365) which is both accessible (access=yes) and navigable(motor_vehicle=yes) gets flagged as sink island because it  is not connected to another navigable road(both the connected trunk roads do not have motor_vehicle or other transportation tags). So after going through the FPs and discussing with @kkonishi2, we have decided the following should determine the accessibility and navigability for this check:
1) **If** access tag is present, its value should be yes (reason being https://github.com/sayas01/atlas-checks/pull/12#discussion_r299255106)
2) **If** any of the transportation tag is present, its value should be yes

Could you all take a look at the updated navigability criteria and share your thoughts on it?

**Review by nachtm — APPROVED** [review]
On the condition that:
 - We're going to do research into proper navigability as far as OSM is concerned and circle back to this with the correct rules.
 - We're very confident that there are no FPs introduced by this change

File: src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java [hunk 10]
```diff
@@ -254,28 +267,65 @@ private boolean isWithinAreasWithExcludedAmenityTags(final Edge edge)
...
+                                area -> Validators.hasValuesFor(area, AmenityTag.class)
+                                        || BuildingTag.isBuilding(area)
+                                        || Validators.hasValuesFor(area, AerowayTag.class))
+                        .spliterator(), false)
+                .anyMatch(area -> area.asPolygon().overlaps(edge.asPolyLine()));
+    }
+
+    /**
+     * Checks if the edge is car navigable in terms of {@link MotorVehicleTag}, {@link MotorcarTag}
+     * and {@link VehicleTag}. Edge is navigable if 1) MotorVehicleTag, MotorcarTag and VehicleTag
```
Comment (seancoulter) [inline src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java:303]: So we're not getting any FP for edges that have an absent  MotorVehicleTag, MotorcarTag and VehicleTag ?

Comment (sayas01) [reply src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SinkIslandCheck.java:303]: @kkonishi2 is doing a final round of analysis on the pre and post enhancement to be 100% sure about this.

**Review by seancoulter — APPROVED** [review]
Thanks for staying on top of this. I think it's in a good state now. I would rather have fewer FP even if that means picking up some FN, but I do think we should revisit this problem once we have clearer navigability criteria.

**Review by danielduhh — CHANGES_REQUESTED** [review]
Approval pending diff metrics from @kkonishi2

**Review by danielduhh — APPROVED** [review]
Summary stats are being re-worked but it sounds like everyone is confident in the results
